### ISSUE TITLE
Roslyn ignores `[MemberNotNull]` annotations on methods in generic classes

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#43749]
**Version Used**: 
```
commit aa691a174b9c809dd210fc43aea8982d0ee579d6
Merge: 995e2628339 81dbb931a2a
Author: Sam Harwell <sam.harwell@microsoft.com>
Date:   Mon Apr 27 23:19:58 2020 -0700

    Merge pull request #43715 from sharwell/intervaltree-any

    <F0><9F><90><87> Avoid using an ArrayBuilder for IntervalTree<T>.Any
```

**Steps to Reproduce**:

Compile the following code
```cs
using System.Diagnostics.CodeAnalysis;
class Base<T>
{
  public string? Prop { get; set; }
  [MemberNotNull(nameof(Prop))] public string Init() => Prop = "";
}

class Program
{
  public static void Main()
  {
    Base<int> b = new Base<int>();
    b.Init();
    b.Prop.ToString();
  }
}
```

**Expected Behavior**:
No warnings. `b.Init()` call should make `b.Prop` not nullable according to its annotation.

**Actual Behavior**:
`CS8602: Dereference of a possibly null reference.` is reported for `b.Prop`

**Notes**
If you remove the type parameter from `Base` the warning goes away i.e. the following code doesn't have the warning:
```
using System.Diagnostics.CodeAnalysis;
class Base
{
  public string? Prop { get; set; }
  [MemberNotNull(nameof(Prop))] public string Init() => Prop = """";
}

class Program
{
  public static void Main()
  {
    Base b = new Base();
    b.Init();
    b.Prop.ToString();
  }
}
```

**Comment by jcouv** [comment]
It looks like this was fixed. There is now no warning produced.

```
#nullable enable
using System.Diagnostics.CodeAnalysis;
class Base<T>
{
  public string? Prop { get; set; }
  [MemberNotNull(nameof(Prop))] public string Init() => Prop = "";
}

class Program
{
  public static void Main()
  {
    Base<int> b = new Base<int>();
    b.Init();
    b.Prop.ToString();
  }
}
```
([sharplab](https://sharplab.io/#v2:CYLg1APgxAdgrgGwQQwEYIKYAIMzZgWACgABABixIEYA6AEQEtkBzGAewGcAXBgYw5oBhNsAwBBPAgCeHBhwDcxEgCYsAIWQcMAHgAqAPmIBvYlkoBmSlTIB+LAAUATmwAOWI1mYYu8rFp9YAL6mWADaALIYALaoGI4Acmxc8YgIABR4URhsAGZpTq4AlIUAuhZWFACSMAxcaYVYALz6Ds5ujVgARJ2KRMFESqrUAOzGISrKISZEZmYaWtoMMFwtqE1YMBgA7uqaOksr9b2zWKg01bVHIWZnBS40umwAylyOS8xXM0HEgUA=))

I'll go ahead and close as fixed. Thanks
